### PR TITLE
Forge scarb plugin backend

### LINKED ISSUE #1905: Migrate attributes defined config to new scarb plugins
### Which components does the task require to be changed? (think hard pls)

snforge

### Description

We should try migrating this feature into the new scarb plugin feature, so that we don't have to maintain it in the scarb codebase.

### PR DESCRIPTION
**Pull request by Draggu**
Closes #1905

## Introduced changes

- changes on forge side to execute config runs

Follow-up to #2044

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/forge-runner/src/lib.rs [hunk 1]
```diff
@@ -153,18 +152,28 @@ fn run_with_fuzzing(
...
+                        cairo_lang_sierra::ids::UserTypeId { ref debug_name, .. },
+                    ) = arg.generic_args[0]
+                    {
+                        debug_name.as_ref().unwrap().as_str()
+                    } else {
+                        name
+                    }
+                } else {
+                    name
+                }
```
Comment (Arcticae) [inline crates/forge-runner/src/lib.rs:168]: This extraction looks like it deserves own function

File: crates/forge-runner/src/package_tests/raw.rs
Comment (Arcticae) [inline crates/forge-runner/src/package_tests/raw.rs:1]: A reminder to address todos in this file

File: crates/forge-runner/src/running/config_run.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        &self,
+        _contract_address: starknet_api::core::ContractAddress,
+        _key: starknet_api::state::StorageKey,
+    ) -> blockifier::state::state_api::StateResult<starknet_api::hash::StarkFelt> {
+        unreachable!()
+    }
+}
+
+#[allow(clippy::too_many_lines)]
+pub fn get_config_for_test_case(
```
Comment (Arcticae) [inline crates/forge-runner/src/running/config_run.rs:64]: I would rename to sth like `run_config_pass` to imply actually running the code

File: crates/forge-runner/src/running/with_config.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+                let args = function_args(func, &type_declarations);
+
+                // trick to fix current fuzzer,
+                // it supports only u256 from types bigger than 1 felt
+                // since we have arguments count we need to know how many u256
+                // are there and add this to length, this way we got
+                // correct unmber of felts, this
+                // should be removed with new fuzzer logic so it is not extensible
+                let u256_occurences = args
```
Comment (Arcticae) [inline crates/forge-runner/src/running/with_config.rs:103]:
```suggestion
                let u256_occurrences = args
```

File: crates/forge-runner/src/running/with_config.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    let default_executables = vec![];
+    let executables = test_target_raw
+        .sierra_program
+        .debug_info
+        .executables
+        .get("snforge_internal_test_executable")
+        .unwrap_or(&default_executables);
+
+    Ok(TestTargetWithConfig {
```
Comment (Arcticae) [inline crates/forge-runner/src/running/with_config.rs:67]: This fragment until the end of the file needs to be refactored because it's unreadable and it's hard to tell what's happening here

File: crates/forge/src/run_tests/resolve_config.rs [hunk 5]
```diff
@@ -56,88 +58,109 @@ async fn resolve_fork_config(
...
+        "hash" => {
+            let block_hash = fork_target.block_id_value.parse::<BigInt>()?;
+
+            BlockId::BlockHash(block_hash.into())
+        }
+        "tag" => {
+            if fork_target.block_id_value == "Latest" {
+                BlockId::BlockTag
+            } else {
+                Err(anyhow!(r#"block tag must be "Latest""#))?
```
Comment (Arcticae) [inline crates/forge/src/run_tests/resolve_config.rs:92]:
```suggestion
                Err(anyhow!(r#"only "Latest" block tag is supported"#))?
```

File: crates/forge/src/scarb.rs [hunk 6]
```diff
@@ -36,23 +41,64 @@ pub fn build_contracts_with_scarb(filter: PackagesFilter) -> Result<()> {
...
-    snforge_target_dir_path: &Utf8Path,
-    package_name: &str,
+    target_dir: &Utf8Path,
+    package: &PackageMetadata,
 ) -> Result<Vec<TestTargetRaw>> {
-    let test_artifacts_path =
-        snforge_target_dir_path.join(format!("{package_name}.snforge_sierra.json"));
+    let mut targets = vec![];
+
+    let dedup_targets: HashMap<_, _> = package
```
Comment (Arcticae) [inline crates/forge/src/scarb.rs:58]: I kind of don't know what's happening here really

File: crates/forge-runner/src/running/hints.rs [hunk 7]
```diff
@@ -0,0 +1,34 @@
+use cairo_lang_casm::hints::Hint;
+use cairo_lang_runner::casm_run::hint_to_hint_params;
+use cairo_vm::serde::deserialize_program::HintParams;
+use std::collections::HashMap;
+use universal_sierra_compiler_api::AssembledCairoProgramWithSerde;
+
+pub fn hints_by_repressentation(
```
Comment (Arcticae) [inline crates/forge-runner/src/running/hints.rs:7]:
```suggestion
pub fn hints_by_representation(
```

File: crates/forge-runner/src/running/config_run.rs [hunk 8]
```diff
@@ -83,11 +83,11 @@ pub fn get_config_for_test_case(
         use_kzg_da: true,
     };
     let (entry_code, builtins) = create_entry_code(args, test_details, casm_program);
-    let footer = SierraCasmRunner::create_code_footer();
 
-    let assembled_program = get_assembled_program(casm_program, entry_code, footer);
+    let assembled_program = get_assembled_program(casm_program, entry_code);
 
-    let (string_to_hint, hints_dict) = create_hints_dict(&assembled_program);
+    let string_to_hint = hints_by_repressentation(&assembled_program);
```
Comment (Arcticae) [inline crates/forge-runner/src/running/config_run.rs:89]:
```suggestion
    let string_to_hint = hints_by_representation(&assembled_program);
```

File: crates/forge-runner/src/running/config_run.rs [hunk 9]
```diff
@@ -1,8 +1,9 @@
-use super::helpers::{create_entry_code, get_assembled_program, run_with_runner};
-use crate::{
-    package_tests::TestDetails,
-    running::{build_syscall_handler, create_hints_dict, get_syscall_segment_index},
+use super::{
+    casm::{get_assembled_program, run_assembled_program},
+    entry_code::create_entry_code,
+    hints::{hints_by_repressentation, hints_to_params},
```
Comment (Arcticae) [inline crates/forge-runner/src/running/config_run.rs:4]:
```suggestion
    hints::{hints_by_representation, hints_to_params},
```

File: crates/forge-runner/src/running.rs [hunk 10]
```diff
@@ -143,11 +146,10 @@ pub fn run_test_case(
 
     let (entry_code, builtins) = create_entry_code(args, &case.test_details, casm_program);
 
-    let footer = SierraCasmRunner::create_code_footer();
-
-    let assembled_program = get_assembled_program(casm_program, entry_code, footer);
+    let assembled_program = get_assembled_program(casm_program, entry_code);
 
-    let (string_to_hint, hints_dict) = create_hints_dict(&assembled_program);
+    let string_to_hint = hints_by_repressentation(&assembled_program);
```
Comment (Arcticae) [inline crates/forge-runner/src/running.rs:151]:
```suggestion
    let string_to_hint = hints_by_representation(&assembled_program);
```

File: crates/forge-runner/src/running.rs [hunk 11]
```diff
@@ -23,23 +24,25 @@ use cheatnet::runtime_extensions::forge_runtime_extension::{
     update_top_call_vm_trace, ForgeExtension, ForgeRuntime,
 };
 use cheatnet::state::{BlockInfoReader, CallTrace, CheatnetState, ExtendedStateReader};
-use helpers::{
-    build_syscall_handler, create_entry_code, create_hints_dict, get_assembled_program,
-    get_syscall_segment_index, run_with_runner,
-};
+use entry_code::create_entry_code;
+use hints::{hints_by_repressentation, hints_to_params};
```
Comment (Arcticae) [inline crates/forge-runner/src/running.rs:28]:
```suggestion
use hints::{hints_by_representation, hints_to_params};
```

File: crates/forge-runner/src/lib.rs [hunk 12]
```diff
@@ -149,25 +166,7 @@ fn run_with_fuzzing(
...
-                        debug_name.as_ref().unwrap().as_str()
-                    } else {
-                        name
-                    }
-                } else {
-                    name
-                }
-            })
-            .collect::<Vec<_>>();
+        let args = args.iter().map(argument_type_name).collect::<Vec<_>>();
```
Comment (Arcticae) [inline crates/forge-runner/src/lib.rs:169]: Those are argument types now

**Review by Arcticae — APPROVED** [review]

File: crates/forge/src/scarb.rs [hunk 13]
```diff
@@ -36,23 +41,70 @@ pub fn build_contracts_with_scarb(filter: PackagesFilter) -> Result<()> {
...
+                TestTargetLocation::Lib
+            } else {
+                TestTargetLocation::Tests
+            };
+
+        let target_file = format!("{target_name}.test.sierra.json");
+
+        match read_to_string(target_dir.join(target_file)) {
+            Ok(value) => {
+                let sierra_program = serde_json::from_str::<ProgramArtifact>(&value)?;
```
Comment (piotmag769) [inline crates/forge/src/scarb.rs:93]: It is a `VersionedProgram` in fact, it works by accident since `VersionedProgram` serializes to 
```
{
  "version": 1,
  // fields of `ProgramArtifact` bcs it is flattened during serialization
} 
```

File: crates/forge/src/scarb.rs [hunk 14]
```diff
@@ -36,23 +41,70 @@ pub fn build_contracts_with_scarb(filter: PackagesFilter) -> Result<()> {
...
+        let tests_location =
+            if target.params.get("test-type").and_then(|v| v.as_str()) == Some("unit") {
+                TestTargetLocation::Lib
+            } else {
+                TestTargetLocation::Tests
+            };
+
+        let target_file = format!("{target_name}.test.sierra.json");
+
+        match read_to_string(target_dir.join(target_file)) {
```
Comment (piotmag769) [inline crates/forge/src/scarb.rs:91]: Context: `cairo-profiler` expects paths to `VersionedProgram` (sierra format) in its input file. Right now `snforge` saves relevant sierra programs to separate files when using `--save-trace-data` or `--build-profile` and then puts paths to these files in the input file for `cairo-profiler`. This is a leftover from using `scarb-snforge-test-collector`
 
With this PR and Scarb 2.7.0 the sierra files with compiled tests are already present on the disk in the form `cairo-profiler` expects. Can we make `snforge` reuse the paths to these files when building input files for `cairo-profiler`? @ksew1 @cptartur 

Can be done in a separate PR

Relevant code:
- saving sierra program
https://github.com/foundry-rs/starknet-foundry/blob/4a1aea0f9f888c9b4774af897fc010d029d0180c/crates/forge/src/run_tests/test_target.rs#L42

- passing the path to the saved sierra program around:
https://github.com/foundry-rs/starknet-foundry/blob/4a1aea0f9f888c9b4774af897fc010d029d0180c/crates/forge-runner/src/running.rs#L85
to later use it here when generating input file for cairo-profiler
https://github.com/foundry-rs/starknet-foundry/blob/4a1aea0f9f888c9b4774af897fc010d029d0180c/crates/forge-runner/src/build_trace_data.rs#L41

Comment (Arcticae) [reply crates/forge/src/scarb.rs:91]: @ksew1 will you create an issue ?🙏

Comment (ksew1) [reply crates/forge/src/scarb.rs:91]: On it

File: crates/forge-runner/src/running/config_run.rs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+use cairo_vm::vm::runners::cairo_runner::ExecutionResources;
+use cheatnet::runtime_extensions::forge_config_extension::{
+    config::RawForgeConfig, ForgeConfigExtension,
+};
+use runtime::{starknet::context::build_context, ExtendedRuntime, StarknetRuntime};
+use starknet_api::block::{BlockNumber, BlockTimestamp};
+use std::{default::Default, num::NonZeroU128};
+use universal_sierra_compiler_api::AssembledProgramWithDebugInfo;
+
+struct FakeStateReader;
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/forge-runner/src/running/config_run.rs:25]: Done in  https://github.com/foundry-rs/starknet-foundry/pull/2264

File: crates/forge/src/scarb.rs (hunk 13 above)
Comment (ksew1) [reply crates/forge/src/scarb.rs:93]: Done in https://github.com/foundry-rs/starknet-foundry/pull/2264

File: crates/forge/src/scarb.rs (hunk 14 above)
Comment (ksew1) [reply crates/forge/src/scarb.rs:91]: https://github.com/foundry-rs/starknet-foundry/issues/2355

File: crates/forge-runner/src/running/with_config.rs (hunk 4 above)
Comment (cptartur) [reply crates/forge-runner/src/running/with_config.rs:78]: Why was this marked as resolved if the file was not changed?

File: crates/forge/src/scarb.rs [hunk 16]
```diff
@@ -36,23 +41,70 @@ pub fn build_contracts_with_scarb(filter: PackagesFilter) -> Result<()> {
...
         .run()
         .context("Failed to build test artifacts with Scarb")?;
     Ok(())
 }
 
+/// collecting by name allow us to dedup targets
+/// we do it because they use same sierra and we display them without distinction anyway
+fn test_targets_by_name(package: &PackageMetadata) -> HashMap<String, &TargetMetadata> {
+    fn test_target_name(target: &TargetMetadata) -> String {
+        // this is logic copied from scarb
```
Comment (cptartur) [inline crates/forge/src/scarb.rs:56]: Would make sense to link relevant file

File: crates/forge/src/scarb.rs [hunk 17]
```diff
@@ -36,23 +41,70 @@ pub fn build_contracts_with_scarb(filter: PackagesFilter) -> Result<()> {
...
+        .arg("--test")
         .packages_filter(filter)
         .run()
         .context("Failed to build test artifacts with Scarb")?;
     Ok(())
 }
 
+/// collecting by name allow us to dedup targets
+/// we do it because they use same sierra and we display them without distinction anyway
+fn test_targets_by_name(package: &PackageMetadata) -> HashMap<String, &TargetMetadata> {
```
Comment (cptartur) [inline crates/forge/src/scarb.rs:54]: So we group targets by name. But in what cases can the target be duplicated if we haven't done this?

File: crates/forge/src/scarb/config.rs [hunk 18]
```diff
@@ -31,24 +30,21 @@ pub struct ForgeConfigFromScarb {
...
-    #[must_use]
-    pub fn params(&self) -> &RawForkParams {
-        &self.params
+    pub fn new(name: String, url: String, block_id_type: String, block_id_value: String) -> Self {
+        Self {
+            name,
+            url,
+            block_id_type,
+            block_id_value,
+        }
```
Comment (cptartur) [inline crates/forge/src/scarb/config.rs:47]: What's preventing the developers from instantiating `ForkTarget` directly now that all it's fields are public? I think we could at least add `#[non_exhaustive]` to `ForkTarget`, so it cannot be created directly.

File: crates/forge/src/test_filter.rs [hunk 19]
```diff
@@ -124,25 +124,25 @@ impl TestCaseFilter for TestsFilter {
...
+
+    fn program_for_testing() -> ProgramArtifact {
+        ProgramArtifact {
+            program: Program {
+                type_declarations: vec![],
+                libfunc_declarations: vec![],
+                statements: vec![],
+                funcs: vec![],
             },
+            debug_info: DebugInfo::default(),
```
Comment (cptartur) [inline crates/forge/src/test_filter.rs:145]: Why default instead of None now?

File: crates/forge/tests/data/forking/src/lib.cairo [hunk 20]
```diff
@@ -91,7 +88,7 @@ mod tests {
     }
 
     #[test]
-    #[fork(url: "{{ NODE_RPC_URL }}", block_id: BlockId::Tag(Latest))]
+    #[fork(url: "{{ NODE_RPC_URL }}", block_tag: latest)]
```
Comment (cptartur) [inline crates/forge/tests/data/forking/src/lib.cairo:91]: In runner code we have an error in code that says that only accepted tag is `Latest` and here we use lowercase. What's the actual error that is displayed to the user if tag other than `latest` is use?

File: crates/forge/tests/e2e/color.rs [hunk 21]
```diff
@@ -26,6 +26,7 @@ fn color_always() {
 }
 
 #[test]
+#[ignore] // TODO: Fix this test, compiling of snforge_scarb_plugin is causing issues, but only in CI
```
Comment (cptartur) [inline crates/forge/tests/e2e/color.rs:29]: TODO

File: crates/forge/tests/e2e/fuzzing.rs [hunk 22]
```diff
@@ -6,7 +6,7 @@ use shared::test_utils::output_assert::{assert_stderr_contains, assert_stdout_co
 fn fuzzing() {
     let temp = setup_package("fuzzing");
 
-    let output = test_runner(&temp).arg("fuzzing").assert().code(1);
+    let output = test_runner(&temp).arg("fuzzing::").assert().code(1);
```
Comment (cptartur) [inline crates/forge/tests/e2e/fuzzing.rs:9]: Why was this changed? Do we really want to run `fuzzing` module only?

File: crates/forge/src/scarb/config.rs (hunk 18 above)
Comment (piotmag769) [reply crates/forge/src/scarb/config.rs:47]: non exhaustive has no effect btw https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute

File: crates/forge/src/scarb.rs (hunk 17 above)
Comment (piotmag769) [reply crates/forge/src/scarb.rs:54]: For integration tests Scarb marks compilation units and targets by group-id, CUs with the same group-id are compiled together. This is done to optimize the compilation (the virtual `lib.cairo` trick I did ~a year ago in our test collector). Simple example: a package with `tests.cairo` and `test_contract.cairo` files in `tests` folder

```json  
      {
          "kind": "test",
          "name": "hello_world_test",
          "source_path": ".../uwu/tests/test.cairo",
          "params": {
            "group-id": "hello_world_integrationtest",
            "test-type": "integration"
          }
        },
        {
          "kind": "test",
          "name": "hello_world_test_contract",
          "source_path": ".../uwu/tests/test_contract.cairo",
          "params": {
            "group-id": "hello_world_integrationtest",
            "test-type": "integration"
          }
        },
```

File: crates/forge/tests/data/forking/src/lib.cairo (hunk 20 above)
Comment (Arcticae) [reply crates/forge/tests/data/forking/src/lib.cairo:91]: Can you link it?

File: crates/forge/tests/e2e/color.rs (hunk 21 above)
Comment (Arcticae) [reply crates/forge/tests/e2e/color.rs:29]: The plugin PR has it as well

File: crates/forge/tests/e2e/steps.rs [hunk 23]
```diff
@@ -21,33 +21,33 @@ fn should_allow_less_than_default() {
 
                 Collected 4 test(s) from steps package
                 Running 4 test(s) from src/
-                [FAIL] steps::tests::steps_4000006
+                [FAIL] steps::tests::steps_4000005
```
Comment (Arcticae) [inline crates/forge/tests/e2e/steps.rs:24]: [nit]: If those change we need to start putting it in the changelog

File: docs/src/snforge-advanced-features/fork-testing.md [hunk 24]
```diff
@@ -53,7 +56,7 @@ the same fork in tests.
 [[tool.snforge.fork]]
 name = "SOME_NAME"
 url = "http://your.rpc.url"
-block_id.tag = "Latest"
+block_id.tag = "latest"
```
Comment (Arcticae) [inline docs/src/snforge-advanced-features/fork-testing.md:59]: It's a breaking change isn't it?

Comment (Arcticae) [reply docs/src/snforge-advanced-features/fork-testing.md:59]: We need to have it in the changelog if it's breaking

File: crates/forge/src/scarb.rs (hunk 16 above)
Comment (Arcticae) [reply crates/forge/src/scarb.rs:56]: Probably will be outdated before a new js framework is released

File: crates/forge-runner/src/package_tests/raw.rs [hunk 25]
```diff
@@ -1,77 +1,31 @@
...
-    pub fuzzer_config: Option<RawFuzzerConfig>,
-    pub test_details: TestDetails,
-}
-
-/// produced by scarb
-#[derive(Debug, Clone, PartialEq, Deserialize)]
-pub enum RawForkConfig {
-    Id(String),
-    Params(RawForkParams),
+//TODO this should be cairo_lang_sierra::program::ProgramArtifact but can't bump cairo_lang_sierra dependency
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/forge-runner/src/package_tests/raw.rs:17]: Fix here https://github.com/foundry-rs/starknet-foundry/pull/2264/commits/11725f9b25fcc1e323cb858a2588853adb1910b4

File: crates/forge-runner/src/package_tests/raw.rs [hunk 26]
```diff
@@ -1,77 +1,31 @@
...
-    pub block_id_type: String,
-    pub block_id_value: String,
-}
-
-/// produced by scarb
-#[derive(Debug, Clone, PartialEq, Deserialize)]
-pub struct RawFuzzerConfig {
-    pub fuzzer_runs: NonZeroU32,
-    pub fuzzer_seed: u64,
+//TODO this should be cairo_lang_sierra::debug_info::DebugInfo but can't bump cairo_lang_sierra dependency
```
[the remark this answers is not in the document]
Comment (ksew1) [reply crates/forge-runner/src/package_tests/raw.rs:25]: Fix here https://github.com/foundry-rs/starknet-foundry/pull/2264/commits/11725f9b25fcc1e323cb858a2588853adb1910b4

File: crates/forge/src/scarb/config.rs (hunk 18 above)
Comment (ksew1) [reply crates/forge/src/scarb/config.rs:47]: Is it worth to add then?

File: crates/forge/tests/e2e/fuzzing.rs (hunk 22 above)
Comment (ksew1) [reply crates/forge/tests/e2e/fuzzing.rs:9]: Yep or else we run  `fuzzing_integrationtest`

File: crates/forge/tests/data/forking/src/lib.cairo (hunk 20 above)
Comment (ksew1) [reply crates/forge/tests/data/forking/src/lib.cairo:91]: This actual error so everything is correct:
```
error: Plugin diagnostic: Both options failed
First variant: #[fork] <block_tag> value incorrect, expected: latest
Second variant: #[fork] can be used with unnamed attributes only
Resolve at least one of them
```

File: docs/src/snforge-advanced-features/fork-testing.md (hunk 24 above)
Comment (ksew1) [reply docs/src/snforge-advanced-features/fork-testing.md:59]: Yep it is, added

File: crates/forge/src/test_filter.rs (hunk 19 above)
Comment (ksew1) [reply crates/forge/src/test_filter.rs:145]: changed https://github.com/foundry-rs/starknet-foundry/pull/2264

File: crates/forge/src/scarb.rs (hunk 16 above)
Comment (cptartur) [reply crates/forge/src/scarb.rs:56]: You can do permanent links on gh, it links to a specific commit

image

File: crates/forge/src/scarb/config.rs (hunk 18 above)
Comment (cptartur) [reply crates/forge/src/scarb/config.rs:47]: It does not have an effect only within defining crate, with others it works. I think we can add it, it doesn't require much effort. But you can add an issue for that and add a todo for future.

File: crates/forge/tests/data/forking/src/lib.cairo (hunk 20 above)
Comment (cptartur) [reply crates/forge/tests/data/forking/src/lib.cairo:91]: Here https://github.com/foundry-rs/starknet-foundry/blob/6e036fc90aeffd026e8a8ef009709ec01ebf8521/crates/forge/src/run_tests/resolve_config.rs#L89

Maybe it's get converted at some point, or it's a different thing and I'm just confused?

Comment (cptartur) [reply crates/forge/tests/data/forking/src/lib.cairo:91]: https://github.com/foundry-rs/starknet-foundry/blob/6e036fc90aeffd026e8a8ef009709ec01ebf8521/crates/forge/src/run_tests/resolve_config.rs#L92

Comment (ksew1) [reply crates/forge/tests/data/forking/src/lib.cairo:91]: Well this is different as it read configuration form `Scarb.toml` , but i agree it should be `latest`. What's weird it is fixed in docs. Anyway changed

**Review by cptartur — APPROVED** [review]
After remaining comments are resolved
